### PR TITLE
Improve `cupy.cutensor`

### PR DESCRIPTION
**Pull request by asi1024**
Depends on #3595.
Improve CPU performance.

- `master`
```
contraction         :    CPU:   55.820 us   +/- 6.812 (min:   49.636 / max:   68.460) us     GPU-0:40316.032 us   +/-383.087 (min:39570.110 / max:40646.305) us
elementwise_binary  :    CPU:   31.403 us   +/- 3.674 (min:   27.876 / max:   37.964) us     GPU-0:  472.358 us   +/- 5.129 (min:  464.608 / max:  479.168) us
elementwise_trinary :    CPU:   37.470 us   +/- 5.565 (min:   31.486 / max:   45.147) us     GPU-0:  563.712 us   +/- 5.574 (min:  557.184 / max:  571.680) us
reduction           :    CPU:   39.862 us   +/- 6.660 (min:   35.085 / max:   52.668) us     GPU-0: 1016.749 us   +/- 6.731 (min: 1012.352 / max: 1030.048) us
```
- This PR (`2e654f7`)
```
contraction         :    CPU:   33.880 us   +/- 6.933 (min:   28.368 / max:   47.218) us     GPU-0:39861.286 us   +/-540.859 (min:38781.727 / max:40186.047) us
elementwise_binary  :    CPU:   20.511 us   +/- 2.704 (min:   17.621 / max:   25.388) us     GPU-0:  461.299 us   +/- 5.706 (min:  452.864 / max:  470.816) us
elementwise_trinary :    CPU:   23.350 us   +/- 2.867 (min:   20.951 / max:   28.692) us     GPU-0:  549.933 us   +/- 3.890 (min:  545.856 / max:  557.024) us
reduction           :    CPU:   23.770 us   +/- 4.193 (min:   20.538 / max:   31.900) us     GPU-0: 1011.245 us   +/- 8.741 (min: 1003.680 / max: 1027.936) us
```
Compared with the scripts in this repository: `examples/cutensor/*.py`.

cupy.sum with cuTENSOR accelerator (implemented in _try_reduction_routine) is also improved.
- `master
```
sum                 :    CPU:   57.215 us   +/- 9.320 (min:   55.607 / max:  913.702) us     GPU-0:  151.111 us   +/-10.814 (min:  147.168 / max:  987.008) us
```
- This PR (`ba1118a`)
```
sum                 :    CPU:   18.730 us   +/- 4.922 (min:   17.982 / max:  481.002) us     GPU-0:  114.798 us   +/- 4.809 (min:  111.904 / max:  560.192) us
```
Compared by running the following script with setting `CUPY_ACCELERATORS='cutensor'`.
```py
import cupy
import cupyx.time
x = cupy.arange(2 ** 24, dtype='float32')
perf = cupyx.time.repeat(cupy.sum, (x,))
print(perf)
```

### CODE REVIEW

**Comment by asi1024** [comment]
Now cpu-time overhead issue in cuTENSOR reduction accelerator was resolved.
```
# w/o cuTENSOR
sum                 :    CPU:   18.878 us   +/-23.512 (min:   16.116 / max:  641.480) us     GPU-0:11594.718 us   +/-25.859 (min:11567.744 / max:12233.152) us
# w/ cuTENSOR
sum                 :    CPU:   18.730 us   +/- 4.922 (min:   17.982 / max:  481.002) us     GPU-0:  114.798 us   +/- 4.809 (min:  111.904 / max:  560.192) us
```

**Comment by leofang** [comment]
Looks nice! @asi1024 could you note what makes the performance improvement? Also, 

> Now cpu-time overhead issue in cuTENSOR reduction accelerator was resolved.

By overhead you meant the large CPU time fluctuation? Why did it happen?

**Comment by asi1024** [comment]
@leofang This PR reduces the accessing of object attribute by annotating cython types.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by asi1024** [comment]
Hmm, I couldn't reproduce the build failure locally. (Perhaps it depends on Cython's version?) I made a fix.
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

File: cupy/cutensor.pyx [hunk 1]
```diff
@@ -2,46 +2,72 @@ import numpy
...
 
     def __repr__(self):
-        return 'mode([' + ', '.join(self.mode) + '])'
+        return self._array.item()
 
 
-def get_handle():
-    dev = device.get_device_id()
+cpdef Handle get_handle():
+    cdef int dev = device.get_device_id()
```
Comment (leofang) [inline cupy/cutensor.pyx:62]: I think you can define `cdef Handle handle` here to make it cdef'd.

File: cupy/cutensor.pyx [hunk 2]
```diff
@@ -2,46 +2,72 @@ import numpy
...
     if dev not in _handles:
-        handle = cutensor.Handle()
+        handle = Handle()
         cutensor.init(handle)
         _handles[dev] = handle
     return _handles[dev]
 
 
-def get_cuda_dtype(numpy_dtype):
+cpdef int get_cuda_dtype(numpy_dtype):
```
Comment (leofang) [inline cupy/cutensor.pyx:70]: I wanted to add `except -1` here, but I think this function will be removed in favor of the change in #3719, so it's fine to not touch it.

File: cupy/cutensor.pyx [hunk 3]
```diff
@@ -56,7 +82,7 @@ def get_cuda_dtype(numpy_dtype):
         raise TypeError('Dtype {} is not supported'.format(numpy_dtype))
 
 
-def get_cutensor_dtype(numpy_dtype):
+cpdef int get_cutensor_dtype(numpy_dtype):
```
Comment (leofang) [inline cupy/cutensor.pyx:85]:
```suggestion
cpdef int get_cutensor_dtype(numpy_dtype) except -1:
```

File: cupy/cutensor.pyx [hunk 4]
```diff
@@ -89,7 +115,7 @@ def create_mode(*mode):
     return Mode(integer_mode)
 
 
-def _auto_create_mode(array, mode):
+cdef inline Mode _auto_create_mode(ndarray array, mode):
```
Comment (leofang) [inline cupy/cutensor.pyx:118]:
```suggestion
cdef inline Mode _auto_create_mode(ndarray array, mode) except -1:
```

File: cupy/cutensor.pyx (hunk 1 above)
Comment (asi1024) [reply cupy/cutensor.pyx:62]: This interface has been exposed to the users, so it should be defined by `cpdef` for backward compatibility.

Comment (asi1024) [reply cupy/cutensor.pyx:62]: However, I don't think the users want this interface, so we can drop `get_handle`, `get_cuda_dtype` and `get_cutensor_dtype` in v8. How do you think? (cc: @kmaehashi)

File: cupy/cutensor.pyx (hunk 4 above)
Comment (asi1024) [reply cupy/cutensor.pyx:118]: We don’t need to declare exception values for functions that return a Python object.
https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values

File: cupy/cutensor.pyx (hunk 3 above)
Comment (asi1024) [reply cupy/cutensor.pyx]: Fixed in 1b16b96. Thanks!

File: cupy/cutensor.pyx (hunk 2 above)
Comment (asi1024) [reply cupy/cutensor.pyx]: Fixed in 1b16b96. Thanks!

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

File: cupy/cutensor.pyx (hunk 4 above)
Comment (leofang) [reply cupy/cutensor.pyx:118]: Yeah I know. I could be wrong, but I remember Cython is being weird in this case and would not propagate the exception properly. Let me check later (on my phone now).

Comment (leofang) [reply cupy/cutensor.pyx:118]: I was wrong, probably had something else in mind, sorry 😅

File: cupy/cutensor.pyx (hunk 1 above)
Comment (kmaehashi) [reply cupy/cutensor.pyx:62]: I think it's ok to break these interfaces.

File: cupy/cutensor.pyx [hunk 5]
```diff
@@ -107,7 +133,8 @@ def _set_compute_dtype(array_dtype, compute_dtype=None):
     return compute_dtype
 
 
-def create_tensor_descriptor(a, uop=cutensor.OP_IDENTITY):
+cpdef TensorDescriptor create_tensor_descriptor(
+        ndarray a, int uop=cutensor.OP_IDENTITY, Handle handle=None):
```
Comment (kmaehashi) [inline cupy/cutensor.pyx:137]: How about making `handle` as a keyword argument and name it `_handle` to make it clear it is a private interface?

File: cupy/cutensor.pyx [hunk 6]
```diff
@@ -186,49 +214,50 @@ def elementwise_trinary(alpha, A, desc_A, mode_A,
...
         compute_dtype = A.dtype
-    alpha = numpy.asarray(alpha, compute_dtype)
-    beta = numpy.asarray(beta, compute_dtype)
-    gamma = numpy.asarray(gamma, compute_dtype)
-    handle = get_handle()
-    cuda_dtype = get_cuda_dtype(compute_dtype)
+
+    return _elementwise_trinary_impl(
+        get_handle(),
+        _Scalar(alpha, compute_dtype), A, desc_A, _auto_create_mode(A, mode_A),
```
Comment (kmaehashi) [inline cupy/cutensor.pyx:234]: Why not `_create_scalar_with_cache`?

File: cupy/cutensor.pyx (hunk 1 above)
Comment (leofang) [reply cupy/cutensor.pyx:62]: Sorry, this is what I meant:
```python
cpdef Handle get_handle():
    cdef Handle handle
    # and the rest follows
```

File: cupy/cutensor.pyx (hunk 2 above)
Comment (leofang) [reply cupy/cutensor.pyx]: I think you can cimport and reuse this function:
https://github.com/cupy/cupy/blob/c3afff0e34fab6325c3db77da442ca5d24882b1c/cupy/core/_dtype.pxd#L3

File: cupy/cutensor.pyx (hunk 6 above)
Comment (kmaehashi) [reply cupy/cutensor.pyx:234]: Ok, `alpha` may be a numpy ndarray which is not hashable.

File: cupy/cutensor.pyx (hunk 2 above)
Comment (asi1024) [reply cupy/cutensor.pyx]: `to_cuda_dtype` works only for `numpy.dtype` object. It does not work if `compute_dtype` of str type is given. (e.g., `cutensor.reduction(..., compute_dtype='d')`)

File: cupy/cutensor.pyx (hunk 5 above)
Comment (asi1024) [reply cupy/cutensor.pyx:139]: We can't use a keyword-only argument for `cpdef` or `cdef` function.

File: cupy/cutensor.pyx (hunk 1 above)
Comment (asi1024) [reply cupy/cutensor.pyx:63]: Fixed! Thanks!

File: cupy/cutensor.pyx (hunk 2 above)
Comment (leofang) [reply cupy/cutensor.pyx]: 1. I skimmed over and don't think you have passed str types to `get_cuda_dtype`
2. I still think it's best to reuse the same function whenever possible. If a str type argument is needed, how about revising `to_cuda_dtype` in `_dtype.pyx` this way?
```python
    if dtype in ('f', numpy.float32):
        return runtime.CUDA_R_32F
    ...
```

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Pull request merged**